networkx/algorithms/cycles.py

```diff
@@ -463,3 +465,157 @@ def find_cycle(G, source=None, orientation='original'):
 
     return cycle[i:]
 
+
+@not_implemented_for('directed')
+@not_implemented_for('multigraph')
+def minimum_cycle_basis(G, weight=None):
+    """
+    Returns a minimum weight cycle basis for G
+    It is defined as a cycle basis for which the total weight (length 
+    for unweighted graphs) of all the cycles is minimum. 
+
+    Parameters
+    ----------
+    G : NetworkX Graph
+    weight: edge attribute to use
+
+    Returns
+    -------
+    A list of cycle lists.  Each cycle list is a list of nodes
+    which forms a cycle (loop) in G. Note that the nodes are not
+    necessarily returned in a order by which they appear in the cycle
+
+    Examples
+    --------
+    >>> G=nx.Graph()
+    >>> G.add_cycle([0,1,2,3])
+    >>> G.add_cycle([0,3,4,5])
+    >>> print(nx.minimum_cycle_basis(G))
+    [[0, 1, 2, 3], [0, 3, 4, 5]]
+
+    References:    
+        [1] Kavitha, Telikepalli, et al. "An O(m^2n) Algorithm for Minimum Cycle Basis of Graphs."
+        http://link.springer.com/article/10.1007/s00453-007-9064-z
+        [2] de Pina, J. 1995. Applications of shortest path methods. Ph.D. thesis, University of Amsterdam, Netherlands 
+
+    See Also
+    --------
+    simple_cycles, cycle_basis
+
+    """
+
+    # We first split the graph in commected subgraphs
+    return sum((_min_cycle_basis(c, weight) for c in
+     nx.connected_component_subgraphs(G)), [])
+
+def _min_cycle_basis(comp, weight):
+    cb = []
+    # We  extract the edges not in this spanning tree
+    spanning_tree_edges = list(
+        nx.minimum_spanning_edges(comp, weight=None, data=False))
+    edges_excl = [
+        frozenset(e) for e in comp.edges() if e not in spanning_tree_edges]
+    N = len(edges_excl)
+
+    # We maintain a set of vectors orthogonal to sofar found cycles
+    set_orth = [set([edge]) for edge in edges_excl]
+    for k in range(N):
+        # kth cycle is "parallel" to kth vector in set_orth
+        new_cycle = _min_cycle(comp, set_orth[k], weight=weight)
+        cb.append(list(set().union(*new_cycle)))
+        # now update set_orth so that k+1,k+2... th elements are
+        # orthogonal to the newly found cycle
+        set_orth = _update_orth(set_orth, comp, new_cycle, k)
+
+    return cb
+
+
+def _update_orth(set_orth, graph, new_cycle, k):
+    """
+    Updates the set of vectors orthogonal to the first k basis cycles,
+    as per [p. 336, 1]
+    """
+    for orth in set_orth[k + 1:]:
```

Comment:
This might also be expressed more clearly as

    base = set_orth[k]
    return [orth ^ base for orth in set_orth[k + 1:] if len(orth & new_cycle) % 2]

This would also lead me to suggest that this simple one-line code should be placed directly in the main loop, not in a separate function that is called in only one other place.